src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -533,6 +533,23 @@ void G1BarrierSetC2::late_barrier_analysis() const {
   compute_liveness_at_stubs();
 }
 
+int G1BarrierSetC2::estimate_stub_size() const {
+  Compile* const C = Compile::current();
+  BufferBlob* const blob = C->output()->scratch_buffer_blob();
+  GrowableArray<G1BarrierStubC2*>* const stubs = barrier_set_state()->stubs();
+  int size = 0;
+
+  for (int i = 0; i < stubs->length(); i++) {
+    CodeBuffer cb(blob->content_begin(), checked_cast<CodeBuffer::csize_t>((address)C->output()->scratch_locs_memory() - blob->content_begin()));
+    MacroAssembler masm(&cb);
+    stubs->at(i)->emit_code(masm);
+    size += cb.insts_size();
+  }
+
+  // Add slop to avoid expansion during emit_stubs.
+  return size + PhaseOutput::MAX_inst_size;
```

Comment:
Is this needed? `init_buffer` already bumps `code_req` up by `MAX_inst_size` immediately after calling `bs->estimate_stub_size()` and that headroom is maintained at each instruction emit.

Reply:
Yes. That slop is going to be taken by normal code itself. This particular addition matches `maybe_expand_to_ensure_remaining(PhaseOutput::MAX_inst_size)` in `G1BarrierSetC2::emit_stubs` just below it. In other words, `emit_stubs` would always maintain `MAX_inst_size` remaining capacity, and that would happen after the normal code is emitted. So we need this slop addition here.

Reply:
Ok, good to go then.